File: problems/binary_trees/find_nodes_distance_k/cpp/find_nodes_distance_k.cpp
#include "find_nodes_distance_k.h"
#include <functional>
#include <queue>
#include <unordered_map>
#include <utility>

// function to find all nodes at distance K from the target node
std::vector<int> findNodesDistanceK(BinaryTree *tree, int target, int k) {

  if (!tree)
    return {}; // if the tree is empty, return an empty vector

  // map to store the parent of each node
  std::unordered_map<BinaryTree *, BinaryTree *> parentMap;

  std::function<void(BinaryTree *, BinaryTree *)> findParents =
      [&](BinaryTree *node, BinaryTree *parent) -> void {
    if (!node)
      return;

    parentMap[node] = parent;
    findParents(node->left, node);
    findParents(node->right, node);
  };

  findParents(tree, nullptr);

  std::function<BinaryTree *(BinaryTree *, int)> findTarget =
      [&](BinaryTree *node, int target) -> BinaryTree * {
    if (!node)
      return nullptr;
    if (node->value == target)
      return node;
    BinaryTree *leftResult = findTarget(node->left, target);
    return leftResult ? leftResult : findTarget(node->right, target);
  };

  // find the target node
  BinaryTree *targetNode = findTarget(tree, target);
  if (!targetNode)
    return {}; // if target node is not found, return an empty vector

  // map to track visited nodes and avoid cycles
  std::unordered_map<BinaryTree *, bool> visited;
  // queue to perform BFS, storing each node and its distance from the target
  std::queue<std::pair<BinaryTree *, int>> q;
  q.push({targetNode, 0});
  visited[targetNode] = true;

  // vecto to store the result
  std::vector<int> result;

  // lambda to add nodes to the queue if unvisited
  auto addToQueue = [&](BinaryTree *node, int currentDistance) -> void {
    if (node && !visited[node]) {
      visited[node] = true;
      q.push({node, currentDistance + 1});
    }
  };

  // BFS
  while (!q.empty()) {
    auto [currentNode, currentDistance] = q.front();
    q.pop();

    // if the current distance equals k, add the node's value to the result
    if (currentDistance == k)
      result.push_back(currentNode->value);

    addToQueue(currentNode->left, currentDistance);
    addToQueue(currentNode->right, currentDistance);
    addToQueue(parentMap[currentNode], currentDistance);
  }

  return result;
}

File: problems/binary_trees/find_nodes_distance_k/python/find_nodes_distance_k.py
from problems.binary_trees.assets.BinaryTree import BinaryTree


def findNodesDistanceK(tree: BinaryTree, target: int, k: int) -> list[int]:

    def find_parents(
        node: BinaryTree, parent_map: dict = None, parent: BinaryTree = None
    ) -> dict[BinaryTree, BinaryTree]:
        if not parent_map:
            parent_map = {}

        if not node:
            return parent_map

        parent_map[node] = parent
        find_parents(node.left, parent_map, node)
        find_parents(node.right, parent_map, node)
        return parent_map

    def find_target(node: BinaryTree, target_value: int) -> BinaryTree:
        if not node:
            return None
        if node.value == target_value:
            return node

        left_result = find_target(node.left, target_value)
        if left_result:
            return left_result
        return find_target(node.right, target_value)

    if not tree or k < 0:
        return []

    parent_map = find_parents(tree)

    target_node = find_target(tree, target)
    if not target_node:
        return []

    visited = set()
    queue = [(target_node, 0)]
    visited.add(target_node)

    result = []

    while queue:
        current_node, current_distance = queue.pop(0)

        if current_distance == k:
            result.append(current_node.value)

        def add_to_queue(node: BinaryTree) -> None:
            if node and node not in visited:
                visited.add(node)
                queue.append((node, current_distance + 1))

        add_to_queue(current_node.left)
        add_to_queue(current_node.right)
        add_to_queue(parent_map.get(current_node))

    return result
